R/test-threads.R

```diff
@@ -0,0 +1,141 @@
+                                            overwrite = TRUE)) %>%
+                 add_tags(paste("test",.x,"threads")))
+
+  mod_paths <- lapply(.mods, function(mod.x){mod.x$absolute_model_path}) %>% unlist()
+
+  submit_models(.mods, .mode = .mode, .wait = FALSE)
+
+  .mods
+}
+
+#' Check estimation time for models run with various threads values
+#'
+#' @param .mods list of bbi model objects created by `test_threads()`.
+#' @param return_times character vector denoting which times from `model_summary()` you want to return.
+#'        See details for more information.
+#' @param wait logical. If TRUE, pass `.mods` to `wait_for_nonmem()` before returning results.
+#' @param ... args passed through to `wait_for_nonmem()`.
+#'
+#' @details
+#' `return_times` can be any subset of `c("estimation_time", "covariance_time", "cpu_time")`.
+#' Users can also specify `all`, which is the shorthand method for selecting all 3 of those columns
+#'
+#' @examples
+#' \dontrun{
+#' mods <- test_threads(mod, threads = c(2, 4))
+#'
+#' If models have not finished:
+#' check_threads(mods, wait = TRUE, time_limit = 300)
+#' check_threads(mods, wait = TRUE, return_times = c("estimation_time", "covariance_time"))
+#'
+#' If models have already finished:
+#' check_threads(mods, wait = FALSE)
+#' check_threads(mods, wait = FALSE, return_times = "All")
+#' }
+#' @return A tibble with columns `threads` (number of threads) and `time`
+#'   (elapsed estimation time in seconds for test models).
+#'
+#' @importFrom tidyselect all_of
+#'
+#' @export
+check_threads <- function(
+  .mods,
+  return_times = "estimation_time",
+  wait = TRUE,
+  ...
+) {
+
+  if("all" %in% return_times){
+    return_times <- c("estimation_time", "covariance_time", "cpu_time")
+  }else{
+    assert_true(all(return_times %in% c("estimation_time", "covariance_time", "cpu_time")))
+  }
+
+  tryCatch({
+    if(wait) wait_for_nonmem(.mods, ...)
+
+    purrr::map_dfr(.mods, ~ {
+      s <- model_summary(.x)
+      threads <- as.numeric(.x$bbi_args$threads)
+      tibble::tibble(threads = threads,
+                     estimation_time = s$run_details$estimation_time,
+                     covariance_time = s$run_details$covariance_time,
+                     cpu_time = s$run_details$cpu_time) %>%
+        select(threads, all_of(return_times))
+    })
+  }, error = function(cond){
+    return(NA)
+  }, warning = function(cond){
+    message(cond)
+    message("\nConsider setting/increasing the `time_limit` in `wait_for_nonmem()`. See ?check_threads for details")
+    return(NA)
+  }
+  )
+}
+
+#' Remove model files associated with the specified tags
+#'
+#' @param .mods a bbi model object or list of model objects.
+#' @param .tags a character vector identifying the tags of the models you want to delete.
+#'              If set to NULL, delete all associated model files.
+#'
+#' @importFrom checkmate check_character
+#'
+#' @export
+cleanup_mods <- function(.mods, .tags = NULL){
```

Comment:
Yeah I went through the same thought process in terms of not being able to pipe it anymore. I also couldn't think of an alternative unless `check_threads()` _also_ returned the model objects, which seemed odd to do. 

I also had the same thought process regarding the default. The original default was "test threads", but `grepl` wouldn't work because I made the tags `test 2 threads`, `test 4 threads`, etc (even without `fixed=TRUE` it wouldn't work because of the number). I could get rid of the number, which would make that much easier, but wasn't sure if we wanted to do that. I agree with removing `NULL` as the default, and only made it that so users were very aware that was possible. I also imagined users would likely _not_ pipe in a bunch of models, and only want to delete some, but what do I know about that.